src/Razor/src/Microsoft.AspNetCore.Razor.LanguageServer/Refactoring/RazorComponentRenameEndpoint.cs

```diff
@@ -0,0 +1,292 @@
+            }
+
+            return new WorkspaceEdit
+            {
+                DocumentChanges = documentChanges,
+            };
+        }
+
+        public void AddFileRenameForComponent(List<WorkspaceEditDocumentChange> documentChanges, DocumentSnapshot documentSnapshot, string newPath)
+        {
+            var oldUri = new UriBuilder
+            {
+                Path = documentSnapshot.FilePath,
+                Host = string.Empty,
+                Scheme = Uri.UriSchemeFile,
+            }.Uri;
+            var newUri = new UriBuilder
+            {
+                Path = newPath,
+                Host = string.Empty,
+                Scheme = Uri.UriSchemeFile,
+            }.Uri;
+
+            documentChanges.Add(new WorkspaceEditDocumentChange(new RenameFile
+            {
+                OldUri = oldUri.ToString(),
+                NewUri = newUri.ToString(),
+            }));
+        }
+
+        private static string MakeNewPath(string originalPath, string newName)
+        {
+            var newFileName = $"{newName}{Path.GetExtension(originalPath)}";
+            var newPath = Path.Combine(Path.GetDirectoryName(originalPath), newFileName);
+            return newPath;
+        }
+
+        public async Task AddEditsForCodeDocument(List<WorkspaceEditDocumentChange> documentChanges, TagHelperBinding originTagHelperBinding, string newName, string documentPath, CancellationToken cancellationToken)
+        {
+            var documentSnapshot = await Task.Factory.StartNew(() =>
+            {
+                _documentResolver.TryResolveDocument(documentPath, out var documentSnapshot);
+                return documentSnapshot;
+            }, cancellationToken, TaskCreationOptions.None, _foregroundDispatcher.ForegroundScheduler).ConfigureAwait(false);
+
+            if (documentSnapshot is null)
+            {
+                return;
+            }
+
+            var codeDocument = await documentSnapshot.GetGeneratedOutputAsync().ConfigureAwait(false);
+            if (codeDocument.IsUnsupported())
+            {
+                return;
+            }
+
+            if (!FileKinds.IsComponent(codeDocument.GetFileKind()))
+            {
+                return;
+            }
+
+            var uri = new UriBuilder
+            {
+                Path = documentPath,
+                Host = string.Empty,
+                Scheme = Uri.UriSchemeFile,
+            }.Uri;
+            AddEditsForCodeDocument(documentChanges, originTagHelperBinding, newName, uri, codeDocument);
+        }
+
+        public void AddEditsForCodeDocument(List<WorkspaceEditDocumentChange> documentChanges, TagHelperBinding originTagHelperBinding, string newName, Uri uri, RazorCodeDocument codeDocument)
+        {
+            var documentIdentifier = new VersionedTextDocumentIdentifier { Uri = uri };
+            foreach (var node in codeDocument.GetSyntaxTree().Root.DescendantNodes().Where(n => n.Kind == SyntaxKind.MarkupTagHelperElement))
```

Comment:
Personally I would move this to a variable if only so I could give it a descriptive name.